guidance/models/_model.py

```diff
@@ -268,288 +509,124 @@ def __call__(self, parser, grammar, ensure_bos_token=True):
-                    # if token_pos == len(sampled_token) and not parser.matched(): # not we don't check if we have matched, because then we can generate anything afterwards
-                    #     if _check_dominated(node, parser, self._token_trie.match_version, parser.next_byte_mask()):
-                    #         token_pos = -1
+                    grammar_temp = self._parser.next_byte_temperature()
+                    current_temp = grammar_temp if grammar_temp >= 0 else 0
+                    logits_state = (self._token_ids, self._parser.bytes[self._start_pos:self._forced_pos], current_temp)
+                    break
 
-                    if token_pos > 0:
-                        break # we found a valid token
+        return is_done, logits_state, response_state
 
-                    if parser.matched():
-                        break # if we already have a full match we don't try more tokens we just give up as soon as the model deviates from the grammar
-            
-            # if we just collapased a hidden commit point then we start over looking for a new token
-            if retry_token_gen:
-                continue
-
-            # emit whatever we know will not be hidden
-            new_bytes = parser.bytes[generated_pos:parser.earliest_hidden_start()]
+    def __call__(self, parser, grammar, ensure_bos_token=True):
+        '''Returns a new updated parser state executed through the grammar.
+        
+        Parameters
+        ----------
+        parser : str or Parser
+            This is represents the current state of a guidance parser that will be extended
+            using the passed grammar. If a string is given then we assume the previous parser
+            state is just a fixed string prompt, if a full Parser is given then we extend that
+            parser by appending the new grammar to the parser's current grammar and then
+            inferencing the model. (TODO: implement full parser extension support)
+        grammar: Grammar
+            This is the grammar we are extending the parser with.
+        '''
 
-            # if we cannot consume any more tokens then we are done
-            if not is_forced and token_pos < len(sampled_token) and trie == self._token_trie:
+        self.start(parser, grammar, ensure_bos_token)
 
-                # which if can't consume any more tokens, but we are not yet done
-                if not parser.matched():
-                    parser.matched()
-                    raise self._report_failed_match(trimmed_prompt_prefix + parser.bytes)
-                
-                # TODO: if we exactly match the end of the pattern then we can commit to this last token 
-                # if m.span()[1] == len(generated_text):
-                #     self._cache_state["new_token_ids"].append(sampled_token_ind)
+        logits = None
+        while True:
+            is_done, logits_state, response_state = self.next(logits)
```

Comment:
I think currently we return each token that is forced as a separate chunk, this in theory allows us to report to the client the probability of each token. It looks like this might make each forced region of bytes into a single chunk is that right? (not the huge issue, but one to note).

Reply:
Unless I introduced a bug, it should return the same number of token responses.